third_party/nvidia/include/Dialect/NVWS/IR/NVWSOps.td

```diff
@@ -134,4 +134,44 @@ def NVWS_ArefReturnOp: NVWS_Op<"aref.return",
   let hasVerifier = 1;
 }
 
+def NVWS_WarpGroupOp : NVWS_Op<"warp_group", [
+  RecursiveMemoryEffects, RecursivelySpeculatable,
+]> {
+  let summary = "Container Op for Warp Specialization";
+  let description = [{
+    Higher level container for Warp Specialization Analysis.
+
+    Contains a variadic number warp groups, a startWarp number, and
+    the number of warps in each group, plus a region to hold the
+    computation for that warp group.
+
+    Regions are not Isolated from Above to aid in analysis,
+    and take inputs purely by reference.
+
+    nvws.warp_group should be lowered to ttg.warp_specialize
+    before execution.
+  }];
+
+  let arguments = (ins
+    I32Attr:$startWarp,
+    NVWS_IntArray:$numWarps);
```

Comment:
why not use DenseI32ArrayAttr? This gives you an `ArrayRef<int32_t>` directly.

Reply:
I was just repeating what i did with the Type Array in the last PR. No strong preference, they appear approximately equivalent.